lib/gemstash/gem_pusher.rb

```diff
@@ -38,12 +39,27 @@ def storage
       @storage ||= Gemstash::Storage.for("private").for("gems")
     end
 
+    def spec_storage
+      @spec_storage ||= Gemstash::Storage.for("private").for("specs")
+    end
+
+    def full_name
+      @full_name ||= gem.spec.full_name
+    end
+
     def check_auth
       Gemstash::Authorization.check(@auth_key, "push")
     end
 
     def store_gem
-      storage.resource(gem.spec.full_name).save(@content, indexed: true)
+      storage.resource(full_name).save(@content, indexed: true)
+    end
+
+    def store_gemspec
+      spec = gem.spec
+      spec = Marshal.dump(spec)
+      spec = Zlib::Deflate.deflate(spec)
+      spec_storage.resource(full_name).save(spec)
     end
 
```

Comment:
:+1: I think we can take a more careful look at refactoring once both our branches are in :-)